docs/features/tagging/high-level-design.md

```diff
@@ -0,0 +1,151 @@
+
+- **All tag managing operations are done per project**: this limits quite a bit the number rows that queries must
+  handle, since doing system wide queries are not contemplated.
+- **System wide features are non-taggable**.
+- **Low volume of feature entries**: each project will have features in the hundreds, low thousands at most.
+- **Low number of tag entries**: the initial requirement is to have a max of 1 tag per feature; it is plausible to
+  assume that even in the case that this limitation is expanded upon, the possible number of tags per feature would be
+  in the low tens.
+- **One tag per feature**: queries are further simplified because there won't be querying for tag combinations
+- **The original `features` table remains the same**
+- **Budget**
+
+## Workflow
+
+There are two distinct parts to the tagging system.
+
+### Feature tagging operations
+
+These endpoints deal with operating tags for an individual feature.
+
+- `PATCH /api/v1/geo-features/:featureId/tags`
+
+```typescript
+Payload
+{
+	projectId: string
+	tag: string;
+}
+```
+
+Changes the current tag with the payload's `tag` for the feature with `featureId`.
+`tag` should not be empty and should be validated to not to exceed the configurable maximum number of characters allowed
+for tag names.
+
+In order to maintain capitalization uniqueness, a check is made first to see if there is already a case insensitive
+match with a previously existing tag for the same Project with `projectId`. If so, that tag will be used instead of the
+tag provided in the payload.
+
+- Extend `POST /api/v1/projects/:id/features/shapefile`
+
+The `UploadShapefileDTO` should be expanded to also accept an optional tag field, so that **all Features created from
+the shape file will be tagged with the same tag**.
+
+It should not be empty and should be validated to not to exceed the configurable maximum number of characters allowed
+for tag names.
+
+In order to maintain capitalization uniqueness, a check is made first to see if there is already a case insensitive
+match with a previously existing tag for the same Project with `projectId`. If so, that tag will be used instead of the
+tag provided in the payload.
+
+- `DELETE /api/v1/geo-features/:featureId/tags/`
+
+Removes the tag for the given `featureId`. This means removing the corresponding row on the `feature_tags` table.
+
+## Tag Managing
+
+These endpoints deal with tag managing operations on a per Project level.
+
+- `GET /api/v1/projects/:projectId/tags?tag=someName`
+
+For the given `projectId`, returns a list of all the DISTINCT tags from the Project's Features that have a tag that
+partially matches with `tag`. By default, the matching is `containing`, but other options could be considered.
+
+Ordered alphabetically by default.
+
+No pagination required, since the possible number of tags per Project will be in the low tens.
+
+- `DELETE /api/v1/projects/:projectId/tags/`
+
+```typescript
+Payload
+{
+	tag: string;
+}
+```
```

Comment:
A payload on a DELETE may be risky: although it should work _in practice_, we don't fully control the whole HTTP stack in production/staging and as such... I would discourage this.

I'd either use the uuid of the tag to delete (which is a bit of an extra pain to get and use, but allows to limit the URL to ASCII) or the name of the tag (maybe urlencoded to avoid having to trust that all layers of the HTTP stack correctly handle unicode in URLs?)

Reply:
For context, see https://stackoverflow.com/a/299696. ChatGPT also invites to be cautious 😬

Reply:
Yeah, thinking back on it, I totally agree.
However, I think using the uuid of the tag wouldn't be correct, because that identifies a single instance of a given tag for that project. If you have multiple features with the same tag, using the uuid it's going to delete the tag for one feature.
Using the tag text as a query parameter, urlencoded as  you say, would most likely be the way to go.

Reply:
of course, you are very right. perfect as in your latest commit 👍🏼